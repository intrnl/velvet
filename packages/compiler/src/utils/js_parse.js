import * as acorn from 'acorn';
import { walk } from 'estree-walker';
import { generate } from 'astring';

/**
 * @param {string} source
 * @returns {import('estree').Program}
 */
export function parse (source) {
	/** @type {import('estree').Comment[]} */
	let comments = [];

	let program = acorn.parse(source, {
		onComment: comments,
		ecmaVersion: 'latest',
		sourceType: 'module',
	});

	reattach_comments(program);
	return program;
}

export function parse_expression (source, position = 0) {
	/** @type {import('estree').Comment[]} */
	let comments = [];

	let node = acorn.parseExpressionAt(source, position, {
		onComment: comments,
		ecmaVersion: 'latest',
		sourceType: 'module',
	});

	reattach_comments(node, comments);
	return node;
}

/**
 * @param {import('estree').Node} ast
 * @param {*} map
 * @returns {string}
 */
export function print (ast, map) {
	return generate(ast, {
		comments: true,
		sourceMap: map,
	});
}

function reattach_comments (ast, comments) {
	walk(ast, {
		/** @param {import('estree').Node} node */
		enter (node) {
			/** @type {import('estree').Comment} */
			let comment;
			let leading_comments = node.leadingComments ||= [];

			while ((comment = comments[0]) && comment.start < node.start) {
				leading_comments.push(comments.shift());
			}
		},
		/** @param {import('estree').Node} node */
		leave (node) {
			let trailing_comments = node.trailingComments ||= [];

			if (comments[0]) {
				let slice = source.slice(node.end, comments[0].start);

				if (/^[,) \t]*$/.test(slice)) {
					trailing_comments.push(comments.shift());
				}
			}
		},
	});
}
