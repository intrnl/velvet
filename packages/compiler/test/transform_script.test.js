import { describe, it, expect } from 'vitest';

import { finalize_program, transform_script } from '../src/transform_script.js';
import { parse, print } from '../src/utils/js_parse.js';


describe('ref', () => {
	it('unused variables', () => {
		let program = parse(`
			let value1;
			let value2 = 100;
			let value3 = new Date();
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('variables with no mutation', () => {
		let program = parse(`
			console.log({ value1, value2, value3, value4 });

			let value1;
			let value2 = 100;
			let value3 = new Date();
			let value4 = value2;

			console.log({ value1, value2, value3, value4 });

			function increment () {
				let value4 = 200;

				console.log({ value1, value2, value3, value4 });
			}
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('variables with mutation', () => {
		let program = parse(`
			console.log({ value1, value2, value3, value4 });

			let value1;
			let value2 = 100;
			let value3 = new Date();
			let value4 = value2;

			value1 += 1;
			value2 += 2;
			value3 += 3;
			value4 += 4;

			console.log({ value1, value2, value3, value4 });

			function increment () {
				let value4 = 200;

				value4 += 4;

				console.log({ value1, value2, value3, value4 });
			}
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('unmutated variable referencing mutated variable', () => {
		let program = parse(`
			let value1 = 100;
			let value2 = value1;

			value1 = 200;
			console.log(value1, value2);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('mutated variable, referencing unmutated variable', () => {
		let program = parse(`
			let value1 = 100;
			let value2 = value1;

			value2 = 200;
			console.log(value1, value2);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('variable mutation with logical assignment operators', () => {
		let program = parse(`
			let value1 = 100;

			value1 ??= 200;
			value1 ||= 300;
			value1 &&= 300;
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('unmutated variable accessing member property', () => {
		let program = parse(`
			let state = { count: 0 };
			let current_date = new Date();

			console.log(state.count);
			console.log(current_date.toISOString());
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('mutated variable accessing member property', () => {
		let program = parse(`
			console.log(state.count);
			console.log(current_date.toISOString());

			let state = { count: 0 };
			let current_date = new Date();

			console.log(state.count);
			console.log(current_date.toISOString());

			state = { count: 0 };
			current_date = new Date();
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('unmutated variable mutating member property', () => {
		let program = parse(`
			let state = { count: 0 };

			state.count += 1;
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('mutated variable mutating member property', () => {
		let program = parse(`
			let state = { count: 0 };

			state = { count: 1 };
			console.log(state.count += 1);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});
});

describe('prop', () => {
	it('unused properties', () => {
		let program = parse(`
			export let value1;
			export let value2 = 100;
			export let value3 = new Date();
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('variables with no mutation', () => {
		let program = parse(`
			export let value1;
			export let value2 = 100;
			export let value3 = new Date();
			export let value4 = value2;

			console.log({ value1, value2, value3, value4 });

			function increment () {
				let value4 = 200;

				console.log({ value1, value2, value3, value4 });
			}
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('variables with mutation', () => {
		let program = parse(`
			console.log({ value1, value2, value3, value4 });

			export let value1;
			export let value2 = 100;
			export let value3 = new Date();
			export let value4 = value2;

			value1 += 1;
			value2 += 2;
			value3 += 3;
			value4 += 4;

			console.log({ value1, value2, value3, value4 });

			function increment () {
				let value4 = 200;

				value4 += 4;

				console.log({ value1, value2, value3, value4 });
			}
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('variable referencing unmutated ref', () => {
		let program = parse(`
			let value1 = 1;
			let value2 = () => {};
			export let value3 = value1;
			export let value4 = value2;
			export let value5 = value1 + value2;
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('variable referencing mutated ref', () => {
		let program = parse(`
			let value1 = 1;
			let value2 = new Date();
			export let value3 = value1;
			export let value4 = value2;
			export let value5 = value1 + value2;

			value1 = 3;
			value2 *= 4;
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('handles separate export specifier', () => {
		let program = parse(`
			let _count = 0;
			let foo = 1;

			export { _count as count, foo };
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('throws on exporting default', () => {
		let program = parse(`
			export default 123;
		`);

		expect(() => transform_script(program)).to.throw();
	});
});

describe('computed', () => {
	it('unused variables', () => {
		let program = parse(`
			$: value1;
			$: value2 = 100;
			$: value3 = new Date();
			foo: unrelated = 3;
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('mutated variables', () => {
		let program = parse(`
			$: value1;
			$: value2 = 100;
			$: value3 = new Date();
			foo: unrelated = 2;

			value1 = 1;
			value2 = 2;
			value3 = 3;
			unrelated = 4;
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('variable referencing unmutated ref', () => {
		let program = parse(`
			let value1 = 100;
			let value2 = new Date();
			$: computed1 = value1 * 2;
			$: computed2 = value2;
			foo: unrelated = value1 + value2;

			console.log(value1, computed1);
			console.log(value2, computed2);
			console.log(unrelated);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('variable referencing mutated ref', () => {
		let program = parse(`
			let value1 = 100;
			let value2 = new Date();
			$: computed1 = value1 * 2;
			$: computed2 = value2;
			foo: unrelated = value1 + value2;

			value1 = 200;
			value2 = Date.now();

			console.log(value1, computed1);
			console.log(value2, computed2);
			console.log(unrelated);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('mutating variable referencing unmutated ref', () => {
		let program = parse(`
			let value1 = 100;
			$: computed = value1;
			foo: unrelated = value1;

			computed = 200;
			unrelated = 300;

			console.log(value1, computed);
			console.log(unrelated);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('unmutated variable referencing unmutated ref member property', () => {
		let program = parse(`
			let value1 = { foo: 123 };
			$: computed = value1.foo;
			$: unrelated = value1.foo;

			console.log(value1, computed);
			console.log(unrelated);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('mutated variable referencing unmutated ref member property', () => {
		let program = parse(`
			let value1 = { foo: 123 };
			$: computed = value1.foo;
			foo: unrelated = value1.foo;

			computed = 234;
			unrelated = 345;

			console.log(value1, computed);
			console.log(unrelated);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('unmutated variable referencing mutated ref member property', () => {
		let program = parse(`
			let value1 = { foo: 123 };
			$: computed = value1.foo;
			foo: unrelated = value1.foo;

			value1 = { foo: 234 };

			console.log(value1, computed);
			console.log(unrelated);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('mutated variable referencing mutated ref member property', () => {
		let program = parse(`
			let value1 = { foo: 123 };
			$: computed = value1.foo;
			foo: unrelated = value1.foo;

			value1 = { foo: 234 };
			computed = 345;
			unrelated = 123;

			console.log(value1, computed);
			console.log(unrelated);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});
});

describe('effect', () => {
	it('references unmutated refs', () => {
		let program = parse(`
			$: console.log(value1, value2);
			foo: console.log(value1, value2);

			$: {
				console.log(value1);
				console.log(value2);
			}

			foo: {
				console.log(value1);
				console.log(value2);
			}

			let value1 = { foo: 123 };
			let value2 = 123;

			$: console.log(value1, value2);
			foo: console.log(value1, value2);

			$: {
				console.log(value1);
				console.log(value2);
			}

			foo: {
				console.log(value1);
				console.log(value2);
			}
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('references mutated refs', () => {
		let program = parse(`
			$: console.log(value1, value2);
			foo: console.log(value1, value2);

			$: {
				console.log(value1);
				console.log(value2);
			}

			let value1 = { foo: 123 };
			let value2 = 123;

			value1 = { foo: 234 };
			value2 = 543;

			$: console.log(value1, value2);
			foo: console.log(value1, value2);

			$: {
				console.log(value1);
				console.log(value2);
			}

			foo: {
				console.log(value1);
				console.log(value2);
			}
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});
});

describe('store', () => {
	it('getter', () => {
		let program = parse(`
			$store;
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('multiple getter references', () => {
		let program = parse(`
			function increment () {
				console.log($value1);
			}

			console.log($value2);
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('setter', () => {
		let program = parse(`
			$store = 123;
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('multiple setter references', () => {
		let program = parse(`
			function increment () {
				$value1 += 1;
			}

			$value2 = 2;
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('subscribing to a ref', () => {
		let program = parse(`
			let value1;

			value1 = get_store();
			console.log($value1);
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('retrieving member property', () => {
		let program = parse(`
			console.log($foo.bar);
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('mutating member property', () => {
		let program = parse(`
			$foo.bar = 123;
		`);

		transform_script(program);

		expect(print(program)).toMatchSnapshot();
	});

	it('throws on lone $', () => {
		let program = parse(`
			console.log($);
		`);

		expect(() => transform_script(program)).to.throw();
	});

	it('throws on declaring $ variables', () => {
		let program = parse(`
			let $foo = 'bar';
		`);

		expect(() => transform_script(program)).to.throw();
	});

	it('only alter single $', () => {
		let program = parse(`
			console.log($foo, $$, $$$);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('no affecting inner scope', () => {
		let program = parse(`
			function log () {
				let $bar = 123;
				console.log($bar, $foo);
			}

			console.log($foo);
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});
});

describe('bind', () => {
	it('const and function exports', () => {
		let program = parse(`
			export const magic = 420;

			export let open = false;

			export function toggle () {
				open = !open;
			}
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('separate export specifier', () => {
		let program = parse(`
			const magic = 420;

			export let open = false;

			function toggle () {
				open = !open;
			}

			export { magic as MAGIC_NUMBER, toggle };
		`);

		transform_script(program);

		let result = print(program);
		expect(result).toMatchSnapshot();
	});

	it('only one specifier for a bind', () => {
		let program = parse(`
			export const magic = 420;

			export { magic as MAGIC_NUMBER };
		`);

		expect(() => transform_script(program)).to.throw();
	});
});

describe('reserved', () => {
	it('throws on declaring $$', () => {
		let program = parse(`
			export let $$foo = 'bar';
		`);

		expect(() => transform_script(program)).to.throw();
	});

	it('throws on reassigning $$', () => {
		let program = parse(`
			$$root = null;
		`);

		expect(() => transform_script(program)).to.throw();
	});

	it('can declare $$$ variables', () => {
		let program = parse(`
			export let $$$foo = 123;
			let $$$bar = 333;
		`);

		expect(() => transform_script(program)).to.not.throw();

		let result = print(program);
		expect(result).toMatchSnapshot();
	});
});

describe('final', () => {
	it('props test', () => {
		let program = parse(`
			export let count = -1;
			let magic = 420;

			export { magic as MAGIC };
		`);

		let { props_idx } = transform_script(program);
		finalize_program({ program, props_idx });

		let result = print(program);
		expect(result).toMatchSnapshot();
	});
});
