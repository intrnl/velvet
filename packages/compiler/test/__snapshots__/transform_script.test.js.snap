// Vitest Snapshot v1

exports[`bind > const and function exports 1`] = `
"const magic = 420;
let open = @prop(1, false);
function toggle() {
  open(!open(@access));
}
@bind({
  magic,
  toggle
});
"
`;

exports[`bind > separate export specifier 1`] = `
"const magic = 420;
let open = @prop(0, false);
function toggle() {
  open(!open(@access));
}
@bind({
  MAGIC_NUMBER: magic,
  toggle
});
"
`;

exports[`computed > mutated variable referencing mutated ref member property 1`] = `
"let value1 = @ref({
  foo: 123
});
let computed = @computed(() => value1(@access).foo);
foo: unrelated = value1(@access).foo;
value1({
  foo: 234
});
computed(345);
unrelated = 123;
console.log(value1(@access), computed(@access));
console.log(unrelated);
"
`;

exports[`computed > mutated variable referencing unmutated ref member property 1`] = `
"let value1 = {
  foo: 123
};
let computed = @ref(value1.foo);
foo: unrelated = value1.foo;
computed(234);
unrelated = 345;
console.log(value1, computed(@access));
console.log(unrelated);
"
`;

exports[`computed > mutated variables 1`] = `
"$: value1;
let value2 = @ref(100);
let value3 = @ref(new Date());
foo: unrelated = 2;
value1 = 1;
value2(2);
value3(3);
unrelated = 4;
"
`;

exports[`computed > mutating variable referencing unmutated ref 1`] = `
"let value1 = 100;
let computed = @ref(value1);
foo: unrelated = value1;
computed(200);
unrelated = 300;
console.log(value1, computed(@access));
console.log(unrelated);
"
`;

exports[`computed > unmutated variable referencing mutated ref member property 1`] = `
"let value1 = @ref({
  foo: 123
});
let computed = @computed(() => value1(@access).foo);
foo: unrelated = value1(@access).foo;
value1({
  foo: 234
});
console.log(value1(@access), computed(@access));
console.log(unrelated);
"
`;

exports[`computed > unmutated variable referencing unmutated ref member property 1`] = `
"let value1 = {
  foo: 123
};
let computed = value1.foo;
let unrelated = value1.foo;
console.log(value1, computed);
console.log(unrelated);
"
`;

exports[`computed > unused variables 1`] = `
"$: value1;
let value2 = 100;
let value3 = new Date();
foo: unrelated = 3;
"
`;

exports[`computed > variable referencing mutated ref 1`] = `
"let value1 = @ref(100);
let value2 = @ref(new Date());
let computed1 = @computed(() => value1(@access) * 2);
let computed2 = @computed(() => value2(@access));
foo: unrelated = value1(@access) + value2(@access);
value1(200);
value2(Date.now());
console.log(value1(@access), computed1(@access));
console.log(value2(@access), computed2(@access));
console.log(unrelated);
"
`;

exports[`computed > variable referencing unmutated ref 1`] = `
"let value1 = 100;
let value2 = new Date();
let computed1 = value1 * 2;
let computed2 = value2;
foo: unrelated = value1 + value2;
console.log(value1, computed1);
console.log(value2, computed2);
console.log(unrelated);
"
`;

exports[`effect > references mutated refs 1`] = `
"$: console.log(value1, value2);
foo: console.log(value1, value2);
$: {
  console.log(value1);
  console.log(value2);
}
let value1 = @ref({
  foo: 123
});
let value2 = @ref(123);
value1({
  foo: 234
});
value2(543);
@effect(() => console.log(value1(@access), value2(@access)));
foo: console.log(value1(@access), value2(@access));
@effect(() => {
  console.log(value1(@access));
  console.log(value2(@access));
});
foo: {
  console.log(value1(@access));
  console.log(value2(@access));
}
"
`;

exports[`effect > references unmutated refs 1`] = `
"$: console.log(value1, value2);
foo: console.log(value1, value2);
$: {
  console.log(value1);
  console.log(value2);
}
foo: {
  console.log(value1);
  console.log(value2);
}
let value1 = {
  foo: 123
};
let value2 = 123;
$: console.log(value1, value2);
foo: console.log(value1, value2);
$: {
  console.log(value1);
  console.log(value2);
}
foo: {
  console.log(value1);
  console.log(value2);
}
"
`;

exports[`program finalizer > hoists imports 1`] = `
"import {define} from \\"@intrnl/velvet/internal\\";
import foo from 'foo';
import {bar} from 'bar';
function setup($$root, $$host) {
  console.log(bar);
}
export default define(null, setup, {});
"
`;

exports[`program finalizer > props test 1`] = `
"import {prop} from \\"@intrnl/velvet/internal\\";
let count = prop(0, -1);
let magic = prop(1, 420);
"
`;

exports[`prop > handles separate export specifier 1`] = `
"let _count = @prop(0, 0);
let foo = @prop(1, 1);
"
`;

exports[`prop > unused properties 1`] = `
"let value1 = @prop(0);
let value2 = @prop(1, 100);
let value3 = @prop(2, () => new Date());
"
`;

exports[`prop > variable referencing mutated ref 1`] = `
"let value1 = @ref(1);
let value2 = @ref(new Date());
let value3 = @prop(0, () => value1(@access));
let value4 = @prop(1, () => value2(@access));
let value5 = @prop(2, () => value1(@access) + value2(@access));
value1(3);
value2(value2(@access) * 4);
"
`;

exports[`prop > variable referencing unmutated ref 1`] = `
"let value1 = 1;
let value2 = () => {
};
let value3 = @prop(0, () => value1);
let value4 = @prop(1, () => value2);
let value5 = @prop(2, () => value1 + value2);
"
`;

exports[`prop > variables with mutation 1`] = `
"console.log({
  value1,
  value2,
  value3,
  value4
});
let value1 = @prop(0);
let value2 = @prop(1, 100);
let value3 = @prop(2, () => new Date());
let value4 = @prop(3, () => value2(@access));
value1(value1(@access) + 1);
value2(value2(@access) + 2);
value3(value3(@access) + 3);
value4(value4(@access) + 4);
console.log({
  value1: value1(@access),
  value2: value2(@access),
  value3: value3(@access),
  value4: value4(@access)
});
function increment() {
  let value4 = 200;
  value4 += 4;
  console.log({
    value1: value1(@access),
    value2: value2(@access),
    value3: value3(@access),
    value4
  });
}
"
`;

exports[`prop > variables with no mutation 1`] = `
"let value1 = @prop(0);
let value2 = @prop(1, 100);
let value3 = @prop(2, () => new Date());
let value4 = @prop(3, () => value2(@access));
console.log({
  value1: value1(@access),
  value2: value2(@access),
  value3: value3(@access),
  value4: value4(@access)
});
function increment() {
  let value4 = 200;
  console.log({
    value1: value1(@access),
    value2: value2(@access),
    value3: value3(@access),
    value4
  });
}
"
`;

exports[`ref > mutated variable accessing member property 1`] = `
"console.log(state.count);
console.log(current_date.toISOString());
let state = @ref({
  count: 0
});
let current_date = @ref(new Date());
console.log(state(@access).count);
console.log(current_date(@access).toISOString());
state({
  count: 0
});
current_date(new Date());
"
`;

exports[`ref > mutated variable calling 1`] = `
"let call = @ref(() => 1);
call(() => 2);
call(@access)();
"
`;

exports[`ref > mutated variable mutating member property 1`] = `
"let state = @ref({
  count: 0
});
state({
  count: 1
});
console.log(state(@access).count += 1);
"
`;

exports[`ref > mutated variable, referencing unmutated variable 1`] = `
"let value1 = 100;
let value2 = @ref(value1);
value2(200);
console.log(value1, value2(@access));
"
`;

exports[`ref > postfix 1`] = `
"let count = @ref(0);
count(count(@access) + 1);
count(count(@access) - 1);
"
`;

exports[`ref > unmutated variable accessing member property 1`] = `
"let state = {
  count: 0
};
let current_date = new Date();
console.log(state.count);
console.log(current_date.toISOString());
"
`;

exports[`ref > unmutated variable calling 1`] = `
"let call = () => 1;
call();
"
`;

exports[`ref > unmutated variable mutating member property 1`] = `
"let state = {
  count: 0
};
state.count += 1;
"
`;

exports[`ref > unmutated variable referencing mutated variable 1`] = `
"let value1 = @ref(100);
let value2 = value1(@access);
value1(200);
console.log(value1(@access), value2);
"
`;

exports[`ref > unused variables 1`] = `
"let value1;
let value2 = 100;
let value3 = new Date();
"
`;

exports[`ref > variable mutation with logical assignment operators 1`] = `
"let value1 = @ref(100);
value1(@access) ?? value1(200);
value1(@access) || value1(300);
value1(@access) && value1(300);
"
`;

exports[`ref > variables with mutation 1`] = `
"console.log({
  value1,
  value2,
  value3,
  value4
});
let value1 = @ref();
let value2 = @ref(100);
let value3 = @ref(new Date());
let value4 = @ref(value2(@access));
value1(value1(@access) + 1);
value2(value2(@access) + 2);
value3(value3(@access) + 3);
value4(value4(@access) + 4);
console.log({
  value1: value1(@access),
  value2: value2(@access),
  value3: value3(@access),
  value4: value4(@access)
});
function increment() {
  let value4 = 200;
  value4 += 4;
  console.log({
    value1: value1(@access),
    value2: value2(@access),
    value3: value3(@access),
    value4
  });
}
"
`;

exports[`ref > variables with no mutation 1`] = `
"console.log({
  value1,
  value2,
  value3,
  value4
});
let value1;
let value2 = 100;
let value3 = new Date();
let value4 = value2;
console.log({
  value1,
  value2,
  value3,
  value4
});
function increment() {
  let value4 = 200;
  console.log({
    value1,
    value2,
    value3,
    value4
  });
}
"
`;

exports[`reserved > can declare $$$ variables 1`] = `
"let $$$foo = @prop(0, 123);
let $$$bar = 333;
"
`;

exports[`store > getter 1`] = `
"let $store = @ref();
@cleanup(store.subscribe($store));
$store(@access);
"
`;

exports[`store > multiple getter references 1`] = `
"let $value1 = @ref();
@cleanup(value1.subscribe($value1));
function increment() {
  console.log($value1(@access));
}
let $value2 = @ref();
@cleanup(value2.subscribe($value2));
console.log($value2(@access));
"
`;

exports[`store > multiple setter references 1`] = `
"let $value1 = @ref();
@cleanup(value1.subscribe($value1));
function increment() {
  value1.set($value1(@access) + 1);
}
value2.set(2);
"
`;

exports[`store > mutating member property 1`] = `
"let $foo = @ref();
@cleanup(foo.subscribe($foo));
$foo(@access).bar = 123;
"
`;

exports[`store > no affecting inner scope 1`] = `
"let $foo = @ref();
@cleanup(foo.subscribe($foo));
function log() {
  let $bar = 123;
  console.log($bar, $foo(@access));
}
console.log($foo(@access));
"
`;

exports[`store > only alter single $ 1`] = `
"let $foo = @ref();
@cleanup(foo.subscribe($foo));
console.log($foo(@access), $$, $$$);
"
`;

exports[`store > reference store twice 1`] = `
"let $foo = @ref();
@cleanup(foo.subscribe($foo));
console.log($foo(@access), $foo(@access));
"
`;

exports[`store > retrieving member property 1`] = `
"let $foo = @ref();
@cleanup(foo.subscribe($foo));
console.log($foo(@access).bar);
"
`;

exports[`store > setter 1`] = `
"store.set(123);
"
`;

exports[`store > subscribing to a ref 1`] = `
"let value1 = @ref();
value1(get_store());
let $value1 = @ref();
@cleanup(value1(@access).subscribe($value1));
console.log($value1(@access));
"
`;
